File: src/test/java/org/micoli/php/FileListProcessorTest.java
package org.micoli.php;

import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.testFramework.fixtures.BasePlatformTestCase;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import org.micoli.php.service.FileListProcessor;

public class FileListProcessorTest extends BasePlatformTestCase {

    protected String getTestDataPath() {
        return "src/test/resources/testMarkDownExporterData";
    }

    private void initFixtures(boolean withExclusionRules) {
        if (withExclusionRules) {
            myFixture.addFileToProject(
                    "/.aiignore",
                    """
                    # Ignore test file
                    target/**
                    out/**
                    **/*Test.java
                    """);
        }

        myFixture.addFileToProject("/target/classes/App.class", "compiled content");
        myFixture.addFileToProject("/out/production/Main.class", "compiled content");
        myFixture.addFileToProject("/main/App.java", "source content");
        myFixture.addFileToProject("/main/AppTest.java", "source content");
    }

    public void testBasicFileListProcessor() {
        myFixture.copyDirectoryToProject(".", ".");
        List<VirtualFile> filesToSelect = List.of(new VirtualFile[] {
            myFixture.findFileInTempDir("/"),
        });

        List<VirtualFile> processedFiles = FileListProcessor.findFilesFromSelectedFiles(filesToSelect);

        assertEquals(4, processedFiles.size());
    }

    public void testFileListProcessorWithExclusionRules() {
        initFixtures(true);
        List<VirtualFile> filesToSelect = List.of(new VirtualFile[] {
            myFixture.findFileInTempDir("/target"),
            myFixture.findFileInTempDir("/out"),
            myFixture.findFileInTempDir("/main"),
        });

        List<VirtualFile> fileList = FileListProcessor.findFilesFromSelectedFiles(filesToSelect);
        List<VirtualFile> processedFiles =
                FileListProcessor.filterFiles(myFixture.findFileInTempDir(".aiignore"), fileList);

        assertNotContains(processedFiles, "App.class");
        assertNotContains(processedFiles, "Main.class");
        assertNotContains(processedFiles, "AppTest.java");
        assertContains(processedFiles, "App.java");
    }

    public void testFileListProcessorWithExclusionRulesAndOnlyOneSelection() {
        initFixtures(true);
        List<VirtualFile> filesToSelect = List.of(new VirtualFile[] {
            myFixture.findFileInTempDir("/"),
        });

        List<VirtualFile> fileList = FileListProcessor.findFilesFromSelectedFiles(filesToSelect);
        List<VirtualFile> processedFiles =
                FileListProcessor.filterFiles(myFixture.findFileInTempDir(".aiignore"), fileList);

        assertNotContains(processedFiles, "App.class");
        assertNotContains(processedFiles, "Main.class");
        assertNotContains(processedFiles, "AppTest.java");
        assertContains(processedFiles, "App.java");
    }

    public void testFileListProcessorWithoutExclusionRules() {
        initFixtures(false);
        List<VirtualFile> filesToSelect = List.of(new VirtualFile[] {
            myFixture.findFileInTempDir("/"),
        });

        List<VirtualFile> processedFiles = FileListProcessor.findFilesFromSelectedFiles(filesToSelect);

        assertContains(processedFiles, "App.class");
        assertContains(processedFiles, "Main.class");
        assertContains(processedFiles, "AppTest.java");
        assertContains(processedFiles, "App.java");
    }

    private static void assertContains(List<VirtualFile> processedFiles, String anObject) {
        assertTrue(Arrays.stream(processedFiles.toArray())
                .anyMatch(file -> Objects.requireNonNull(((VirtualFile) file).getCanonicalPath())
                        .endsWith(anObject)));
    }

    private static void assertNotContains(List<VirtualFile> processedFiles, String anObject) {
        assertTrue(Arrays.stream(processedFiles.toArray())
                .noneMatch(file -> Objects.requireNonNull(((VirtualFile) file).getCanonicalPath())
                        .endsWith(anObject)));
    }
}

File: src/test/kotlin/org/micoli/php/FileListProcessorTest.kt
package org.micoli.php

import com.intellij.openapi.vfs.VirtualFile
import com.intellij.testFramework.fixtures.BasePlatformTestCase
import org.assertj.core.api.Assertions.*
import org.micoli.php.service.filesystem.FileListProcessor

class FileListProcessorTest : BasePlatformTestCase() {
    override fun getTestDataPath(): String = "src/test/resources/testMarkDownExporterData"

    private fun initFixtures(withExclusionRules: Boolean) {
        if (withExclusionRules) {
            myFixture.addFileToProject(
                "/.aiignore",
                """
          # Ignore test file
          target/**
          out/**
          **/*Test.java

          """
                    .trimIndent(),
            )
        }

        myFixture.addFileToProject("/target/classes/App.class", "compiled content")
        myFixture.addFileToProject("/out/production/Main.class", "compiled content")
        myFixture.addFileToProject("/main/App.java", "source content")
        myFixture.addFileToProject("/main/AppTest.java", "source content")
    }

    fun testBasicFileListProcessor() {
        myFixture.copyDirectoryToProject(".", ".")
        val filesToSelect = listOf<VirtualFile>(myFixture.findFileInTempDir("/")).toMutableList()

        val processedFiles = FileListProcessor.findFilesFromSelectedFiles(filesToSelect)

        assertThat(processedFiles.size).isEqualTo(4)
    }

    fun testFileListProcessorWithExclusionRules() {
        initFixtures(true)
        val filesToSelect =
            listOf<VirtualFile>(
                    myFixture.findFileInTempDir("/target"),
                    myFixture.findFileInTempDir("/out"),
                    myFixture.findFileInTempDir("/main"),
                )
                .toMutableList()

        val fileList = FileListProcessor.findFilesFromSelectedFiles(filesToSelect).toMutableList()
        val processedFiles =
            FileListProcessor.filterFiles(myFixture.findFileInTempDir(".aiignore"), fileList)

        assertNotContains(processedFiles, "App.class")
        assertNotContains(processedFiles, "Main.class")
        assertNotContains(processedFiles, "AppTest.java")
        assertContains(processedFiles, "App.java")
    }

    fun testFileListProcessorWithExclusionRulesAndOnlyOneSelection() {
        initFixtures(true)
        val filesToSelect = listOf<VirtualFile>(myFixture.findFileInTempDir("/")).toMutableList()

        val fileList = FileListProcessor.findFilesFromSelectedFiles(filesToSelect).toMutableList()
        val processedFiles =
            FileListProcessor.filterFiles(myFixture.findFileInTempDir(".aiignore"), fileList)
                .toMutableList()

        assertNotContains(processedFiles, "App.class")
        assertNotContains(processedFiles, "Main.class")
        assertNotContains(processedFiles, "AppTest.java")
        assertContains(processedFiles, "App.java")
    }

    fun testFileListProcessorWithoutExclusionRules() {
        initFixtures(false)
        val filesToSelect = listOf<VirtualFile>(myFixture.findFileInTempDir("/")).toMutableList()

        val processedFiles =
            FileListProcessor.findFilesFromSelectedFiles(filesToSelect).toMutableList()

        assertContains(processedFiles, "App.class")
        assertContains(processedFiles, "Main.class")
        assertContains(processedFiles, "AppTest.java")
        assertContains(processedFiles, "App.java")
    }

    private fun assertContains(processedFiles: MutableList<VirtualFile>, anObject: String) {
        assertThat(
                processedFiles.stream().anyMatch {
                    ((it as VirtualFile).canonicalPath)?.endsWith(anObject) == true
                })
            .isTrue
    }

    private fun assertNotContains(processedFiles: MutableList<VirtualFile>, anObject: String) {
        assertThat(
                processedFiles.stream().noneMatch {
                    ((it as VirtualFile).canonicalPath)?.endsWith(anObject) == true
                })
            .isTrue
    }
}
